### DISCUSSION TITLE
Custom resource pre serialization handler

### DISCUSSION THREAD

**Discussion by AFE-GmdG** [godotengine/godot-proposals#6365] [Engine Core]
### Describe the project you are working on
I'm working on a RTS maze / labyrinth like game right now. It uses a 3D environment composed of a grid of cells, visible from above.

### Describe the problem or limitation you are having in your project
I'm using custom resources which are inherited from ArrayMesh which is used to generate reusable meshes with additional properties like Custom Vector data:
```py
	@warning_ignore("int_as_enum_without_match")
	var format: Mesh.ArrayFormat = (
		Mesh.ARRAY_FORMAT_VERTEX
		| Mesh.ARRAY_FORMAT_NORMAL
		| Mesh.ARRAY_FORMAT_TEX_UV
		| Mesh.ARRAY_FORMAT_TEX_UV2
		| Mesh.ARRAY_FORMAT_INDEX
		| Mesh.ARRAY_FORMAT_CUSTOM0
		| (Mesh.ARRAY_CUSTOM_RGBA_FLOAT << Mesh.ARRAY_FORMAT_CUSTOM0_SHIFT)
	) as Mesh.ArrayFormat

	add_surface_from_arrays(Mesh.PRIMITIVE_TRIANGLES, surface, [], {}, format)
	if (surfaceMaterial != null):
		surface_set_material(0, surfaceMaterial)
```
This works fine as i'm reaching my target with this soultion. Hovever ay using a resouce like this the Scene file become way to big, because the resource serialization store to many data:
```
[gd_scene load_steps=4 format=3 uid="uid://b8b8dqorg2d76"]

[ext_resource type="Material" uid="uid://dugjl37doaws1" path="res://Materials/FloorDisplaceTest.tres" id="1_qi5u1"]
[ext_resource type="Script" path="res://Meshes/FloorGrid/FloorGridMesh.gd" id="2_b6gr0"]

[sub_resource type="ArrayMesh" id="ArrayMesh_qp3q4"]
_surfaces = [{
"aabb": AABB(-8.1, 0, -8.1, 18.9, 1e-05, 18.9),
"attribute_data": PackedByteArray(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 253, 255, 255, 255, 253, 255, 255, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 253, 255, 255, 255, 253, 255, 255, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 0, 0, 253, 255, 255, 255, 253, 255, 255, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 63, 0, 0, 128, 63, 0, 0, 0, 0, 0, 0, 0, 0, 253, 255, 255, 255, 253, 255, 255, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
...

[node name="CustomDataTestScene" type="Node3D"]
metadata/_edit_lock_ = true

[node name="FloorGridMesh_LOD_0" type="MeshInstance3D" parent="."]
mesh = SubResource("ArrayMesh_qp3q4")
metadata/_edit_lock_ = true
```
Since the mesh data is calculated anyway, it should not saved in the tscn file. By using multiple of these calculated mesh resources the tscn file gets really big (over multiple MB)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
If I'm be able to intercept the serialization process of the resource I could remove unnecessary data. This virtual method should be called by the scene serialization while saving the scene file. It should somehow provide a method to change the stored data, remove some of it or add additional data

### If this enhancement will not be used often, can it be worked around with a few lines of script?
I could inherit from `Mesh` instead of `ArrayMesh`, since the additional stored data come from the `ArrayMesh` internal implementation of `_get_property_list()`. Using `_get_property_list()` in my custom resource let me not override settings from `ArrayMesh` so I could not prevent the storing of the unnecessary data this way.
However inheriting from `Mesh` means a lot of additional work for me because I have to deal with the `RenderingServer` directly and have to overwrite a ton of functions that `ArrayMesh` already implement just fine. Additionally this method may not be usable for other resource types which may not have a common virtual ancestor like `Mesh` is.

I hope this all makes sense and my explanations are sufficient.

**Comment by dalexeev** [comment]
There is the `_validate_property()` method, but it is not yet available in the scripting API.
